stdlib/InteractiveUtils/src/macros.jl

```diff
@@ -37,20 +50,59 @@ function recursive_dotcalls!(ex, args, i=1)
     return ex, i
 end
 
+is_interpolated_type(arg) = isexpr(arg, :call, 2) && arg.args[1] === InterpolatedType
+function _typeof_expr(ex)
+    is_interpolated_type(ex) && return esc(ex.args[2])
+    Core.Typeof(ex)
+end
+
+function extract_farg(arg)
+    !is_interpolated_type(arg) && return esc(arg)
+    fT = esc(arg.args[2])
+    :($construct_callable($fT))
+end
+
+function construct_callable(@nospecialize(func::Type))
+    # Support function singleton types such as `(::typeof(f))(args...)`
+    Base.issingletontype(func) && isdefined(func, :instance) && return func.instance
+    # Don't support type annotations otherwise, we don't want to give wrong answers
+    # for callables such as `(::Returns{Int})(args...)` where using `Returns{Int}`
+    # would give us code for the constructor, not for the callable object.
+    throw(ArgumentError("If a function type is explicitly provided, it must be a singleton whose only instance is the callable object"))
+end
+
+typesof_expr(args) = :(typesof($(esc.(args)...)))
+
+function process_type_annotations!(ex::Expr)
+    for (i, subex) in enumerate(ex.args)
+        isa(subex, Expr) || continue
+        if isexpr(subex, :(::), 1)
+            ex.args[i] = :($InterpolatedType($(subex.args[1])))
+        elseif isexpr(subex, :...) && isexpr(subex.args[1], :(::), 1)
+            T = subex.args[1].args[1]
+            ex.args[i] = :(Vararg{$T})
+        else
+            process_type_annotations!(subex)
```

Comment:
Does this need to be recursive? I would think we can only handle `::T` as a direct argument to the function being examined.

Reply:
I initially thought we may want to support tuples/named tuples recursively, such as `@code_typed f(x, (1, (::Float64, 2)))` (where a similar motivation had me introduce `InterpolatedType` to support runtime splatting such as `f(1, (1, ::Float64)...)`.

But on second thought I'd rather keep user expectations simple with regards to the processing it does, so let's make it only support direct arguments.